src/analysis/souffle/analyze_tag_checks_test_helper.dl

```diff
@@ -0,0 +1,87 @@
+#ifndef SRC_ANALYSIS_SOUFFLE_ANALYZE_TAG_CHECKS_TEST_HELPER_DL_
+#define SRC_ANALYSIS_SOUFFLE_ANALYZE_TAG_CHECKS_TEST_HELPER_DL_
+
+#include "taint.dl"
+#include "analyze_tag_checks.dl"
+
+// This is a tag check that does not correspond to the original test's arcs manifest. We want to
+// eventually translate the arcs manifests into facts and check them against our hand-written facts,
+// so distinguishing the extra tags will be important eventually. For now, thought, it just works
+// exactly like checkTagIsPresnet.
+.decl checkTagIsPresentExtra(accessPath: AccessPath, tag: Tag)
+
+checkTagIsPresent(accessPath, tag) :- checkTagIsPresentExtra(accessPath, tag).
+
+.type CheckResult <: symbol
+.decl isValidCheckResult(checkResult: CheckResult)
+
+isValidCheckResult("PASS").
+isValidCheckResult("FAIL").
+
+// The arcs manifests we use as our intermediate representation will indicate that particular
+// `accessPath`s should be checked for the presence or absence of tags. Our tests, however, should
+// not always be in a passing state if all checks pass, nor should they necessarily be in a failing
+// state if some or all checks fail. Depending upon the functionality that the test is testing and
+// how we have structured the dataflow of the test, we expect that certain checks will pass and
+// certain checks will fail.
+//
+// This relation, `tagCheckExpecation`, encodes our expectation about whether the dataflow analysis
+// on the dataflow graph we have provided with edge and claim facts should cause a particular
+// check to pass or fail. A test passes when each check on a particular `accessPath` for a
+// particular `tag` passes or fails exactly when the `tagCheckExpectation` with the same
+// `accessPath` and `tag` value indicates "PASS" or "FAIL" as its `checkResult`, respectively.
+.decl tagCheckExpectation(
+  accessPath: AccessPath,
+  tag: Tag,
+  checkResult: CheckResult)
+
+.type BadCheckDiagnostic <: symbol
+
+// Used to report malformed test cases.
+.decl badTagCheckExpectation(
+  accessPath: AccessPath,
+  tag: Tag,
+  checkResult: CheckResult,
+  diagnostic: BadCheckDiagnostic)
+
+.output badTagCheckExpectation(IO=stdout, delimiter=",")
```

Comment:
It looks like the logic in this file is comparing the actual value of a Datalog predicate against some expected value. Is there a reason we need/want to do this comparison in Datalog? Would it make things easier if whatever predicates we want to test are just output to the souffle-cpp interface and then this comparison of real/expected values is done in C++ code rather than Datalog?

Reply:
I think Datalog is the right place to do something like this, it is quite convenient and concise to match expectations to facts there. That said, I've greatly simplified the code as per @bgogul's recommendation, so this is probably out of date now.